### PR TITLE
build: Add Java SDK params builder pattern

### PR DESCRIPTION
**Pull request by whutomo**
- Added and modified `api.mustache` template to include builder patten implementation across resource api instances.
- Added `api_test.mustache` to adapt with the changes introduce to use builder pattern on the test files
- Added and modified `pojo.mustache` template to include builder patten implementation across the models.
- Added `model_test.mustache` to adapt with the changes introduce to use builder pattern on the test files
- Updated `openapitools.json` java config to include java mustache template directory

An example of `create` message - (/POST messages) endpoint generated
image

Here's an example of the usage from the client perspective when trying to execute "sending a message" operation
image

### CODE REVIEW

File: codegen/mustache/java/pojo.mustache [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+ * {{description}}{{^description}}{{classname}}{{/description}}{{#isDeprecated}}
+ * @deprecated{{/isDeprecated}}
+ */{{#isDeprecated}}
+@Deprecated{{/isDeprecated}}
+{{#swagger1AnnotationLibrary}}
+{{#description}}
+@ApiModel(description = "{{{.}}}")
+{{/description}}
+{{/swagger1AnnotationLibrary}}
+{{#jackson}}
```
Comment (richardfontein) [inline codegen/mustache/java/pojo.mustache:32]: There's a fair bit of stuff in this template not relevant for our codegen, which will add to the maintenance cost and mental overhead of readability. I'd suggest we remove everything we're not using currently, we can add back later if the use-case arises.

Let's sync to walkthrough things we can remove, based on the actual output of our openapi through the codegen.

File: codegen/mustache/java/api.mustache [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+            throw e;
+        }
+        {{/errorObjectType}}
+        {{^errorObjectType}}
+        Type localVarReturnType = new TypeToken<{{{returnType}}}>(){}.getType();
+        return localVarApiClient.execute(localVarCall, localVarReturnType);
+        {{/errorObjectType}}
+        {{/returnType}}
+    }
+    {{/vendorExtensions.x-streaming}}
```
Comment (richardfontein) [inline codegen/mustache/java/api.mustache:244]: As below, let's see what we can remove from this file. This `x-streaming` vendor extension support is a good example of something we can remove.

**Comment by richardfontein** [comment]
Referencing Stripe, they're able to invoke the `builder` method on nested objects within each Params class, for example the `CustomerCreateParams.Adderss.builder` method - this saves an import for our `PUT` endpoints that don't return a resource. Is this something we can easily do?

image

File: codegen/mustache/java/pojo.mustache (hunk 1 above)
Comment (whutomo) [reply codegen/mustache/java/pojo.mustache:32]: I agree, let's have a discussion to identify more examples that might not be necessary and can be removed.

File: codegen/mustache/java/libraries/okhttp-gson/ApiClient.mustache [hunk 3]
```diff
@@ -0,0 +1,1871 @@
...
+                final String regex = "\\/([^\\/]+)\\/?$";
+                final Pattern pattern = Pattern.compile(regex, Pattern.MULTILINE);
+                final Matcher matcher = pattern.matcher(location);
+                try {
+                    matcher.find();
+                    JsonObject jsonObject = JsonParser.parseString(respBody).getAsJsonObject();
+                    jsonObject.addProperty("docId", matcher.group(1));
+                    return JSON.deserialize(jsonObject.toString(), returnType);
+                } catch (Exception e) {
+                    System.out.println("respBody is:" + respBody);
```
Comment (richardfontein) [inline codegen/mustache/java/libraries/okhttp-gson/ApiClient.mustache:1115]: This hidden error message could be confusing for customers without context of why it was done. I'd suggest rewording to:

```
System.out.println("Expected JSON response body was not received. Actual response body was `" + respBody + "`");
```

File: codegen/package.json [hunk 4]
```diff
@@ -8,6 +8,9 @@
   },
   "bugs": "https://github.com/whispir/openapi/issues",
   "scripts": {
+    "generate:java": "yarn generate --generator-key java --custom-generator ./generators/java/target/java-openapi-generator-1.0.0.jar",
```
Comment (richardfontein) [inline codegen/package.json:11]: Per the update to [generate-sdk workflow](https://github.com/whispir/openapi/pull/176/files#diff-dda97367049910868155a384fc513c14c18b3cb6fec946b17c25ae549f9f9887R37), we need to add a `generate:node` script too

File: codegen/generators/java/src/main/java/com/whispir/codegen/JavaCustomClientCodegen.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    public JavaCustomClientCodegen() {
+        super();
+        apiNameSuffix = "Api";
+    }
+
+    public String getName() {
+        return "java-custom";
+    }
+
+    private class CustomSort implements Comparator<CodegenProperty> {
```
Comment (richardfontein) [inline codegen/generators/java/src/main/java/com/whispir/codegen/JavaCustomClientCodegen.java:27]: Could we rename class to something more descriptive? It's not clear what this sort is doing currently.

`ReadOnlySort` could work well

**Review by richardfontein — CHANGES_REQUESTED** [review]
Great work Mark! Massive effort.

File: codegen/mustache/java/pojo.mustache [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+ * {{description}}{{^description}}{{classname}}{{/description}}
+ */
+
+public class {{classname}} {
+  @SerializedName("docId")
+  private String docId;
+  @SerializedName("rawOutput")
+  private String rawOutput;
+  // Additional getters
+  public String getDocId() {
```
Comment (richardfontein) [inline codegen/mustache/java/pojo.mustache:33]: Could we replace all text instances of `docId` here with just `id`? I think this is possible now given that we've removed all `id` attributes from all models. This also aligns with Stripe API.

**Review by richardfontein — CHANGES_REQUESTED** [review]

File: codegen/mustache/java/pojo.mustache (hunk 6 above)
Comment (richardfontein) [reply codegen/mustache/java/pojo.mustache:33]: Hmm, on second thought, removal of the `id` attribute from the response model of each `GET /<resource>` endpoint actually makes it impossible to use the SDK to perform a `GET /<resource>` query then retrieve the full model for the resource using the `id` param.

We may need to revert https://github.com/whispir/openapi/pull/182/files#diff-d910ba2ef878f7db0223a966b81c8b3f3b65027bb39e4431bb05140171eece39L9223. Let's discuss.

**Review by richardfontein — APPROVED** [review]
